Kernel/Graphics/VirtIOGPU/VirtIOGPU.cpp

```diff
@@ -0,0 +1,286 @@
+        // 2. Create BUFFER using VIRTIO_GPU_CMD_RESOURCE_CREATE_2D
+        create_framebuffer();
+        // 3. Attach backing storage using  VIRTIO_GPU_CMD_RESOURCE_ATTACH_BACKING
+        attach_backing_storage();
+        // 4. Use VIRTIO_GPU_CMD_SET_SCANOUT to link the framebuffer to a display scanout.
+        attach_framebuffer_to_scanout();
+        // 5. Render our test pattern
+        draw_test_pattern();
+        // 6. Use VIRTIO_GPU_CMD_TRANSFER_TO_HOST_2D to update the host resource from guest memory.
+        transfer_framebuffer_data_to_host(m_display_info.rect);
+        // 7. Use VIRTIO_GPU_CMD_RESOURCE_FLUSH to flush the updated resource to the display.
+        flush_displayed_image(m_display_info.rect);
+    } else {
+        VERIFY_NOT_REACHED();
+    }
+}
+
+VirtIOGPU::~VirtIOGPU()
+{
+}
+
+bool VirtIOGPU::handle_device_config_change()
+{
+    return false;
+}
+
+void VirtIOGPU::handle_queue_update(u16 queue_index)
+{
+    dbgln_if(VIRTIO_DEBUG, "VirtIOGPU: Handle queue update");
+    VERIFY(queue_index == CONTROLQ);
+
+    auto& queue = get_queue(CONTROLQ);
+    ScopedSpinLock queue_lock(queue.lock());
+    queue.discard_used_buffers();
+    m_has_outstanding_request = false;
+}
+
+u32 VirtIOGPU::get_pending_events()
+{
+    return config_read32(*m_device_configuration, DEVICE_EVENTS_READ);
+}
+
+void VirtIOGPU::clear_pending_events(u32 event_bitmask)
+{
+    config_write32(*m_device_configuration, DEVICE_EVENTS_READ, event_bitmask);
+}
+
+void VirtIOGPU::query_display_information()
+{
+    Locker locker(m_operation_lock);
+    auto& request = *reinterpret_cast<VirtIOGPUCtrlHeader*>(m_scratch_space->vaddr().as_ptr());
+    auto& response = *reinterpret_cast<VirtIOGPURespDisplayInfo*>((m_scratch_space->vaddr().offset(sizeof(request)).as_ptr()));
+
+    populate_virtio_gpu_request_header(request, VirtIOGPUCtrlType::VIRTIO_GPU_CMD_GET_DISPLAY_INFO, VIRTIO_GPU_FLAG_FENCE);
+
+    synchronous_virtio_gpu_command(start_of_scratch_space(), sizeof(request), sizeof(response));
+
+    for (size_t i = 0; i < VIRTIO_GPU_MAX_SCANOUTS; ++i) {
+        auto& scanout = response.scanout_modes[i];
+        if (!scanout.enabled)
+            continue;
+        dbgln("Scanout {}: x: {}, y: {}, width: {}, height: {}", i, scanout.rect.x, scanout.rect.y, scanout.rect.width, scanout.rect.height);
+        m_display_info = scanout;
+        m_chosen_scanout = i;
+    }
+    m_display_info.rect.width = 1024;
+    m_display_info.rect.height = 768;
+}
+
+void VirtIOGPU::create_framebuffer()
+{
+    Locker locker(m_operation_lock);
+    auto& request = *reinterpret_cast<VirtIOGPUResourceCreate2D*>(m_scratch_space->vaddr().as_ptr());
+    auto& response = *reinterpret_cast<VirtIOGPUCtrlHeader*>((m_scratch_space->vaddr().offset(sizeof(request)).as_ptr()));
+
+    populate_virtio_gpu_request_header(request.header, VirtIOGPUCtrlType::VIRTIO_GPU_CMD_RESOURCE_CREATE_2D, VIRTIO_GPU_FLAG_FENCE);
+
+    request.resource_id = m_framebuffer_id = 1;
```

Comment:
Can we do these separately? we dont use that syntax in serenity usually.

Reply:
Sure, I'll fix this in the next revision.